global route: now consistent with detailed routing w.r.t. congestion failures

oharboe opened the pull request:
Just like detailed routing, a global route that fails with congestion, is not a build failure(as in exit code non-zero), it is a successful(as in zero exit code) global route that produce reports detailing the problem.

Detailed route will not proceed, if there is global routing congestion

This allows build systems, such as bazel, to create artifacts for global and detailed route, even if the operation had problems, without having know about the semantics between global and detailed route.

Considering that global and detailed route can run for a long time and use a lot of memory, this allows inspecting results on a laptop for a build that ran on a server.

oharboe committed:
global route: now consistent with detailed routing w.r.t. congestion failures

Just like detailed routing, a global route that fails with congestion, is not
a build failure(as in exit code non-zero), it is a successful(as in zero exit code)
global route that produce reports detailing the problem.

Detailed route will not proceed, if there is global routing congestion

This allows build systems, such as bazel, to create artifacts for global
and detailed route, even if the operation had problems, without having
know about the semantics between global and detailed route.

Considering that global and detailed route can run for a long time and
use a lot of memory, this allows inspecting results on a laptop for
a build that ran on a server.

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

oharboe commented on flow/scripts/detail_route.tcl:
```diff
@@ -1,4 +1,10 @@
 utl::set_metrics_stage "detailedroute__{}"
+if {[expr [file exists $::env(REPORTS_DIR)/congestion.rpt] && \
+    [file size $::env(REPORTS_DIR)/congestion.rpt] != 0]} {
+  error "Global routing failed, run `make gui_grt` and load $::env(REPORTS_DIR)/congestion.rpt \
```
Detailed route does not proceed if global route had congestion failures.

maliberty replied:
This moves the error to the wrong step.  We already have
```
set result [catch {do_global_route} errMsg]
if {$result != 0} {
  write_db $::env(RESULTS_DIR)/5_1_grt-failed.odb
  error $errMsg
}
```
Why isn't this sufficient?

oharboe replied:
Separation of concerns and also trying to minimize surprises.

I want the responsiblity for providing .odb files and .rpt in the case of global/detailed routing failures and the detection of fatal(no artifacts should be produced) to reside inside ORFS.

The interface I want between ORFS and bazel is that ORFS provides .odb and .rpt file and a non-zero exit code, in which case artifacts are created. With a non-zero exit code, then no artifacts are created. A non-zero exit code is ORFS's way to tell bazel that there was a fatal error and no artifacts should be created.

This mirrors detailed routing's behavior today, I want the same for global routing.

Yes: this means that there needs to be a check not to continue detailed routing when there is a global routing error. This is somewhat surprising, so not ideal, but the error message in detailed routing is easily visible and not drowned out.

One could imagine the same for detailed routing; that there's no point in continuing with the final stage if there the DRC errors in detailed routing are unacceptable(though I don't know what "unacceptable" is for detailed routing).

maliberty replied:
That might be good for Bazel but it runs against how make works.  An artifact is a sign of success not failure in make.

oharboe replied:
I dont think so: this is how detailed route works.

maliberty replied:
It is how gpl & dpl work.  drt is different as people often manually fix drcs left by drt.

oharboe replied:
does dpl and other stages have similarly useful reports in case of failure, other than the log?

maliberty replied:
You can use -report_file_name with detailed_place though it looks like we don't currently in the flow.  I'm not sure why it matters.

oharboe replied:
Just trying to understand where there are frequently useful reports on failure. More reports on place failures could be a useful future improvement.

oharboe committed:
global route: add GENERATE_ARTIFACTS_ON_FAILURE variable

the default is the current Makefile behavior, but ORFS also
supports the bazel use-case where artifacts are needed, even in
the case of e.g. a global route failure.

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

maliberty commented on flow/scripts/fillcell.tcl:
```diff
@@ -1,4 +1,8 @@
-utl::set_metrics_stage "globalroute__{}"
+if {[expr [file exists $::env(REPORTS_DIR)/congestion.rpt] && \
```
FYI, It is likely that this step will get moved after drt in the future.

oharboe replied:
Yes... this support is going to require some maintenance as the flow changes, no way around that. Hopefully not too bad and perhaps we can find some nice way.

maliberty commented on flow/scripts/global_route.tcl:
```diff
@@ -1,120 +1,143 @@
-  if {[info exist ::env(DETAILED_METRICS)]} {
-    report_metrics 5 "global route pre repair design"
-  }
+  write_guides $::env(RESULTS_DIR)/route.guide
```
Depending on the nature of the error this may not make sense.  It would be better to do it only when you don't write a -failed.odb

oharboe replied:
How is the user-experience diminished from having this information in the case of a failed global route?

maliberty replied:
The same reason we write -failed.odb to indicate it isn't valid data.  Putting invalid data in a same place as valid data is confusing at best.

oharboe committed:
Merge remote-tracking branch 'origin/master' into HEAD

oharboe committed:
global route: revert some changes to what it is doing on master

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>
